### PR TITLE
fix(package): Recognize and normalize `cargo.toml` 

### LINKED ISSUE #12384: `cargo publish` with `cargo.toml` on windows doesn't normalize the content
### Problem

On case insensitive filesystems (which can include Windows and Mac), cargo allows `cargo.toml` to be used.  When publishing, the case doesn't get normalize and we miss it when normalizing the content of the file (`prepare_for_publish`).

Once published
- crates.io won't be able to process the file
- the crate is unusable on systems with case-sensitive filesystems
- behavior might be off or flat out broken without normalization (e.g. workspace inheritance)

### Steps

```console
$ cargo new foo
$ cd foo
$ cargo package --list --allow-dirty
Cargo.lock
Cargo.toml
Cargo.toml.orig
src/main.rs
$ move Cargo.toml t
$ move t cargo.toml
$ cargo package --list --allow-dirty
Cargo.lock
Cargo.toml
src/main.rs
```

Note the lack of `Cargo.toml.orig`

### Possible Solution(s)

Long term, I'd prefer we error, suggesting a rename if `cargo.toml` exists

To help people transition, we should warn first.

Alternatively
- we could normalize the file name
- we could detect this and error when loading manifests

### Notes

See also
- #11962
- #45
- https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/broken.20crate.20files
- https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/.60cargo.2Etoml.60.20on.20case.20insensitive.20filesystems

### Version

_No response_

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This solution is a blend of conservative and easy
- Normalizes `cargo.toml` to `Cargo.toml` on publish
  - Ensuring we publish the `prepare_for_publish` version and include `Cargo.toml.orig`
  - Avoids dealing with trying to push existing users to `Cargo.toml`
- All other cases of `Cargo.toml` are warnings
  - We could either normalize or turn this into an error in the future
  - When helping users with case previously, we've only handle the `cargo.toml` case
  - We already should have a fallback in case a manifest isn't detected
  - I didn't want to put in too much effort to make the code more complicated to handle this

As a side effect, if a Linux user has `cargo.toml` and `Cargo.toml`, we'll only put one of them in the `.crate` file.  We can extend this out to also include a warning for portability for case insensitive filesystems but I left that for after rust-lang/cargo#12235.

### How should we test and review this PR?

A PR at a time will show how the behavior changed as the source was edited

This does add a direct dependency on `unicase` to help keep case-insensitive comparisons easy / clear and to avoid riskier areas for bugs like writing an appropriate `Hash` implementation.  `unicase` is an existing transitive dependency of cargo.

### Additional information

Fixes #12384

[Discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/.60cargo.2Etoml.60.20on.20case.20insensitive.20filesystems)

### CODE REVIEW

**Review by weihanglo — COMMENTED** [review]
> Another case this does not address (though slightly changes) is when a Linux system has multiple cases of Cargo.toml

Will this cause problems for projects already having both `Cargo.toml` and `cargo.toml`? Should we examine existing packages on crates.io?

Apart from that, looks pretty good!

**Comment by epage** [comment]
> Will this cause problems for projects already having both `Cargo.toml` and `cargo.toml`? Should we examine existing packages on crates.io?

To lazy to examine them so I took an approach for handling it.  Thoughts?

**Comment by ChrisDenton** [comment]
Hm... unicase seems overkill no? IIRC, cargo's special filenames are all ascii.

**Comment by epage** [comment]
> Hm... unicase seems overkill no? IIRC, cargo's special filenames are all ascii.

I wrote this with an eye for portability hazards in the future.  Unsure if filesystems are case insensitive by ascii or by unicode but thought I'd play it safe rather than people in the future building on the this work and not taking this into account.

**Comment by ChrisDenton** [comment]
Sure but I mean, pure ascii filenames are the only ones that can be compared portably.

What version of the Unicode spec does a filesystem use to do case-insensitive comparisons? Trick question! It could be using any version or its own weird case table. So the only portable thing would be to ask the filesystem, unless you stick to ascii only.

**Comment by epage** [comment]
@ChrisDenton to clarify, how much of your concern is with
- `unicase::UniCase`
- `unicase` as a dependency
- any dependency for this

**Comment by ChrisDenton** [comment]
Well I do prefer having fewer dependencies but that's only a minor concern.

And I'm not against using `unicase::UniCase` exactly. It'll work for `Cargo.toml`, but then so would an ascii comparison. My concern is a bit more fundamental in that using any single case algorithm is wrong. With ascii comparisons the limitation is at least clear to the reader and the results correct within its domain.

The "correct" way to do this for arbitrary names outside of ascii is to try opening with the file then comparing the canonical name of the opened file with the expected name. But I realise this might be more involved then is wanted here, especially considering cross-platform concerns.

**Comment by epage** [comment]
Thanks for the clarification.

I've updated to ASCII comparisons.  I kept `unicase` because it makes life easier in a couple of ways, in particular by choosing to use a `HashMap`, it makes it easy to get a correct `Hash` implementation.  It is already a transitive dependency of cargo.

File: tests/testsuite/package.rs [hunk 1]
```diff
@@ -2983,3 +2983,115 @@ src/main.rs.bak
...
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                exclude = ["*.txt"]
+                license = "MIT"
+                description = "foo"
+            "#;
+    let p = project()
+        .file("Cargo.toml", manifest)
+        .file("cargo.toml", manifest)
```
Comment (riking) [inline tests/testsuite/package.rs:3059]: Test should put a bogus manifest into the lowercase version and assert it's not the one chosen

Comment (riking) [reply tests/testsuite/package.rs:3059]: Ah, you stated above you aren't dealing with that in this change.

File: tests/testsuite/package.rs [hunk 2]
```diff
@@ -2983,3 +2983,60 @@ src/main.rs.bak
...
+        p.root().join("cargo.toml"),
+        r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                exclude = ["*.txt"]
+                license = "MIT"
+                description = "foo"
+            "#,
```
Comment (weihanglo) [inline tests/testsuite/package.rs:3006]: nit: One less indentation?

**Review by weihanglo — APPROVED** [review]
Looks good to me, and thanks everybody here for helping review :)

**Pull request merged**